File: packages/openapi/src/contracts/session.ts
import { initContract } from "@ts-rest/core";
import { z } from "zod";
import {
  ZSessionResponse,
  ZSessionListResponse,
  ZCreateSessionBody,
  ZListSessionsQuery,
  ZSessionParams,
} from "@genta/zod";
import { getSecurityMetadata } from "@/utils.js";

const c = initContract();

export const sessionContract = c.router({
  // GET /api/v1/sessions
  listSessions: {
    summary: "List study sessions",
    path: "/api/v1/sessions",
    method: "GET",
    description: "Get paginated list of study sessions for the current user",
    query: ZListSessionsQuery,
    responses: {
      200: ZSessionListResponse,
      401: z.object({ message: z.string() }),
    },
    metadata: getSecurityMetadata(),
  },

  // POST /api/v1/sessions
  createSession: {
    summary: "Start a new study session",
    path: "/api/v1/sessions",
    method: "POST",
    description: "Start a new study session for the current user",
    body: ZCreateSessionBody,
    responses: {
      201: ZSessionResponse,
      400: z.object({ message: z.string() }),
      401: z.object({ message: z.string() }),
    },
    metadata: getSecurityMetadata(),
  },

  // GET /api/v1/sessions/:session_id
  getSession: {
    summary: "Get study session by ID",
    path: "/api/v1/sessions/:session_id",
    method: "GET",
    description: "Get details of a specific study session",
    pathParams: ZSessionParams,
    responses: {
      200: ZSessionResponse,
      401: z.object({ message: z.string() }),
      404: z.object({ message: z.string() }),
    },
    metadata: getSecurityMetadata(),
  },

  // PUT /api/v1/sessions/:session_id/end
  endSession: {
    summary: "End a study session",
    path: "/api/v1/sessions/:session_id/end",
    method: "PUT",
    description: "End a study session and calculate duration/accuracy",
    pathParams: ZSessionParams,
    body: z.object({}).optional(),
    responses: {
      200: ZSessionResponse,
      400: z.object({ message: z.string() }),
      401: z.object({ message: z.string() }),
      404: z.object({ message: z.string() }),
    },
    metadata: getSecurityMetadata(),
  },
});

File: apps/backend/internal/handler/session.go
package handler

import (
	"net/http"

	"github.com/labstack/echo/v4"
	"github.com/manikandareas/genta/internal/middleware"
	"github.com/manikandareas/genta/internal/model"
	"github.com/manikandareas/genta/internal/model/session"
	"github.com/manikandareas/genta/internal/server"
	"github.com/manikandareas/genta/internal/service"
)

type SessionHandler struct {
	Handler
	sessionService *service.SessionService
}

func NewSessionHandler(s *server.Server, sessionService *service.SessionService) *SessionHandler {
	return &SessionHandler{
		Handler:        NewHandler(s),
		sessionService: sessionService,
	}
}

// CreateSession godoc
// @Summary Start a new study session
// @Description Start a new study session for the current user
// @Tags sessions
// @Accept json
// @Produce json
// @Param body body session.CreateSessionRequest true "Session data"
// @Success 201 {object} session.SessionResponse
// @Failure 400 {object} errs.HTTPError
// @Failure 401 {object} errs.HTTPError
// @Router /sessions [post]
func (h *SessionHandler) CreateSession(c echo.Context) error {
	return Handle(
		h.Handler,
		func(c echo.Context, req *session.CreateSessionRequest) (*session.SessionResponse, error) {
			userID := middleware.GetUserID(c)
			return h.sessionService.Create(c, userID, req)
		},
		http.StatusCreated,
		&session.CreateSessionRequest{},
	)(c)
}

// ListSessions godoc
// @Summary List study sessions
// @Description Get paginated list of study sessions for the current user
// @Tags sessions
// @Accept json
// @Produce json
// @Param page query int false "Page number" default(1)
// @Param limit query int false "Items per page" default(10)
// @Success 200 {object} model.PaginatedResponse[session.SessionResponse]
// @Failure 401 {object} errs.HTTPError
// @Router /sessions [get]
func (h *SessionHandler) ListSessions(c echo.Context) error {
	return Handle(
		h.Handler,
		func(c echo.Context, req *session.ListSessionsRequest) (*model.PaginatedResponse[session.SessionResponse], error) {
			userID := middleware.GetUserID(c)
			return h.sessionService.List(c, userID, req)
		},
		http.StatusOK,
		&session.ListSessionsRequest{},
	)(c)
}

// EndSession godoc
// @Summary End a study session
// @Description End a study session and calculate duration/accuracy
// @Tags sessions
// @Accept json
// @Produce json
// @Param session_id path string true "Session ID"
// @Success 200 {object} session.SessionResponse
// @Failure 400 {object} errs.HTTPError
// @Failure 401 {object} errs.HTTPError
// @Failure 404 {object} errs.HTTPError
// @Router /sessions/{session_id}/end [put]
func (h *SessionHandler) EndSession(c echo.Context) error {
	return Handle(
		h.Handler,
		func(c echo.Context, req *session.EndSessionRequest) (*session.SessionResponse, error) {
			userID := middleware.GetUserID(c)
			return h.sessionService.End(c, userID, req.SessionID)
		},
		http.StatusOK,
		&session.EndSessionRequest{},
	)(c)
}

// GetSession godoc
// @Summary Get a study session by ID
// @Description Get details of a specific study session
// @Tags sessions
// @Accept json
// @Produce json
// @Param session_id path string true "Session ID"
// @Success 200 {object} session.SessionResponse
// @Failure 401 {object} errs.HTTPError
// @Failure 404 {object} errs.HTTPError
// @Router /sessions/{session_id} [get]
func (h *SessionHandler) GetSession(c echo.Context) error {
	return Handle(
		h.Handler,
		func(c echo.Context, req *session.GetSessionRequest) (*session.SessionResponse, error) {
			userID := middleware.GetUserID(c)
			return h.sessionService.GetByID(c, userID, req.SessionID)
		},
		http.StatusOK,
		&session.GetSessionRequest{},
	)(c)
}
